File: Samyam-Go/Lab1/PlayFair.go
package main

import (
    "strings"
)

type PlayfairCipher struct {
    key    string
    matrix [5][5]byte
}

func NewPlayfairCipher(key string) *PlayfairCipher {
    cipher := &PlayfairCipher{key: key}
    cipher.generateMatrix()
    return cipher
}

func (p *PlayfairCipher) generateMatrix() {
    k := 0
    used := [26]bool{}

    for _, c := range p.key {
        if !isAlpha(byte(c)) {
            continue
        }
        uc := toUpper(c)
        if !used[uc-'A'] && uc != 'J' {
            p.matrix[k/5][k%5] = byte(uc)
            used[uc-'A'] = true
            k++
        }
    }

    for c := byte('A'); c <= 'Z'; c++ {
        if c == 'J' {
            continue
        }
        if !used[c-'A'] {
            p.matrix[k/5][k%5] = c
            k++
        }
    }
}

func (p *PlayfairCipher) findPosition(letter byte) (int, int) {
    for i := 0; i < 5; i++ {
        for j := 0; j < 5; j++ {
            if p.matrix[i][j] == letter {
                return i, j
            }
        }
    }
    return -1, -1
}

func (p *PlayfairCipher) encryptPair(a, b byte) string {
    if a == b {
        b = 'X'
    }

    posA1, posA2 := p.findPosition(a)
    posB1, posB2 := p.findPosition(b)

    if posA1 == posB1 {
        return string(p.matrix[posA1][(posA2+1)%5]) +
            string(p.matrix[posB1][(posB2+1)%5])
    } else if posA2 == posB2 {
        return string(p.matrix[(posA1+1)%5][posA2]) +
            string(p.matrix[(posB1+1)%5][posB2])
    } else {
        return string(p.matrix[posA1][posB2]) +
            string(p.matrix[posB1][posA2])
    }
}


func (p *PlayfairCipher) encrypt(plaintext string) string {
    encryptedText := ""

    plaintext = removeNonAlpha(plaintext)

    if len(plaintext)%2 != 0 {
        plaintext += "X"
    }

    for i := 0; i < len(plaintext); i += 2 {
        a := toUpper(rune(plaintext[i]))
        b := toUpper(rune(plaintext[i+1]))
        encryptedText += p.encryptPair(byte(a), byte(b))
    }

    return encryptedText
}

func removeNonAlpha(s string) string {
    return strings.Map(func(r rune) rune {
        if isAlpha(byte(r)) {
            return r
        }
        return -1
    }, s)
}

func isAlpha(c byte) bool {
    return 'A' <= c && c <= 'Z'
}

func toUpper(c rune) byte {
    return byte(strings.ToUpper(string(c))[0])
}

File: Samyam-Cpp/Lab1/PlayFair.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

class PlayfairCipher {
private:
    string key;
    char matrix[5][5];

    void generateMatrix() {

        int k = 0;
        bool used[26] = {false};

        // Populate the matrix with the key
        for (char c : key) {
            if (!isalpha(c))
                continue;
            char uc = toupper(c);
            if (!used[uc - 'A'] && uc != 'J') { // Ignore duplicates and 'J'
                matrix[k / 5][k % 5] = uc;
                used[uc - 'A'] = true;
                k++;
            }
        }

        // Fill the remaining matrix with the alphabet (excluding 'J')
        for (char c = 'A'; c <= 'Z'; c++) {
            if (c == 'J') // Skip 'J'
                continue;
            if (!used[c - 'A']) {
                matrix[k / 5][k % 5] = c;
                k++;
            }
        }
    }

    pair<int, int> findPosition(char letter) {
        for (int i = 0; i < 5; i++) 
            for (int j = 0; j < 5; j++) 
                if (matrix[i][j] == letter)
                    return make_pair(i, j);
        return make_pair(-1, -1);
    }

    string encryptPair(char a, char b) {

        if (a == b)
            b = 'X';

        auto posA = findPosition(a);
        auto posB = findPosition(b);

        if (posA.first == posB.first) 
            return string(1, matrix[posA.first][(posA.second + 1) % 5]) +
                   string(1, matrix[posB.first][(posB.second + 1) % 5]);
        
        else if (posA.second == posB.second) 
            return string(1, matrix[(posA.first + 1) % 5][posA.second]) +
                   string(1, matrix[(posB.first + 1) % 5][posB.second]);
        
        else 
            return string(1, matrix[posA.first][posB.second]) +
                   string(1, matrix[posB.first][posA.second]);
        
    }

    string decryptPair(char a, char b) {

        auto posA = findPosition(a);
        auto posB = findPosition(b);

        if (posA.first == posB.first) 
            return string(1, matrix[posA.first][(posA.second + 4) % 5]) +
                   string(1, matrix[posB.first][(posB.second + 4) % 5]);
        else if (posA.second == posB.second) 
            return string(1, matrix[(posA.first + 4) % 5][posA.second]) +
                   string(1, matrix[(posB.first + 4) % 5][posB.second]);
        
        else 
            return string(1, matrix[posA.first][posB.second]) +
                   string(1, matrix[posB.first][posA.second]);
        
    }

public:

    PlayfairCipher(string key) : key(key) {
        generateMatrix();
    }

    string encrypt(string plaintext) {
        string encryptedText;

        plaintext.erase(remove_if(plaintext.begin(), plaintext.end(), [](char c) { return !isalpha(c); }), plaintext.end());

        if (plaintext.length() % 2 != 0)
            plaintext += 'X';

        for (size_t i = 0; i < plaintext.length(); i += 2) {
            char a = toupper(plaintext[i]);
            char b = toupper(plaintext[i + 1]);
            encryptedText += encryptPair(a, b);
        }

        return encryptedText;
    }

    string decrypt(string ciphertext) {
        string decryptedText;

        for (size_t i = 0; i < ciphertext.length(); i += 2) {
            char a = toupper(ciphertext[i]);
            char b = toupper(ciphertext[i + 1]);
            decryptedText += decryptPair(a, b);
        }

        return decryptedText;
    }

    void printMatrix() {
        cout << "Playfair Cipher Matrix:" << endl;
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                cout << matrix[i][j] << " ";
            }
            cout << endl;
        }
    }
};

int main() {
    string key;
    cout << "Enter the key: ";
    cin >> key;
    
    PlayfairCipher cipher(key);
    cipher.printMatrix(); // Print the key matrix
    
    string plaintext;
    cout << "Enter the plaintext: ";
    cin >> plaintext;
    
    string encrypted = cipher.encrypt(plaintext);
    cout << "Encrypted: " << encrypted << endl;
    
    string decrypted = cipher.decrypt(encrypted);
    cout << "Decrypted: " << decrypted << endl;

    cout << "Name: Samyam Subedi \nRoll No: 86 \nSection: C" << endl;
    
    return 0;
}
